### ISSUE TITLE
Cordon, drain and remove a node from the cluster when running kubeadm reset

### ISSUE THREAD

**Issue by mikedanese** [kubernetes/kubeadm#42]
_From @bruceauyeung on October 9, 2016 2:32_

currently `kubeadm` command provides a command `join` to join a cluster. but there isn't a command to leave that cluster. maybe `kubeadm leave` is suitable for this circumstance

kubeadm version

> version.Info{Major:"1", Minor:"5+", GitVersion:"v1.5.0-alpha.0.1534+cf7301f16c0363-dirty", GitCommit:"cf7301f16c036363c4fdcb5d4d0c867720214598", GitTreeState:"dirty", BuildDate:"2016-09-27T18:10:39Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"}

_Copied from original issue: kubernetes/kubernetes#34404_

**Comment by mikedanese** [comment]
_From @errordeveloper on October 13, 2016 9:18_

This makes sense to add. Are you envisioning re-joining another cluster or simplify clean-up and recycle nodes? Right now you can just throw away a node and run `kubectl delete node <name>` to deregister it.

**Comment by mikedanese** [comment]
_From @pires on October 14, 2016 20:0_

@luxas please assign this to me.

**Comment by mikedanese** [comment]
_From @pires on October 14, 2016 20:55_

I'll add this behavior after #34807 lands, such that both
`kubeadm reset` and `kubeadm leave` result in the node been marked as unschedulable to the cluster.

Thoughts @luxas @errordeveloper?

**Comment by mikedanese** [comment]
_From @pires on November 8, 2016 9:22_

This is a list of potential execution steps:
* Mark current node as **unshedulable** _aka_ cordon the node
* [Drain the node](http://kubernetes.io/docs/user-guide/kubectl/kubectl_drain/)
* Restart the `kubelet` with empty or default values, e.g. crash-loop until an API server is provided

1 - Existing `kubectl` code living in `pkg/kubectl/cmd/drain.go` should be either re-used or adapted.
2 - `kubeadm reset` should execute similarly, differing only in the last step where instead of restarting, it should stop the `kubelet`.

Thoughts?

**Comment by pires** [comment]
> Restart the kubelet with empty or default values, e.g. crash-loop until an API server is provided

I raised discussion about how to change `kubelet` flags in #64.

Really need thoughts before proceeding.

**Comment by luxas** [comment]
We need to discuss what the semantics would be regarding `leave` vs `reset`.
Could we just fix `reset` to actually remove the node from the cluster as well (right now the node will be left in `kubectl get no` with status `NotReady`)

What would the use-case for `leave` be that `reset` wouldn't do?

**Comment by pires** [comment]
@luxas _leaving_ a cluster is different than _reset_ everything, including stopping the `kubelet`.

**Comment by luxas** [comment]
@pires so when would you _leave_ but not _reset_? 
That's what I want to know...

Are you thinking about something like a node restart? But that can be fixed with a plain kubectl command...
I just want to figure out the use case for `leave`

**Comment by pires** [comment]
@luxas i joined master 1. now i want my node to join a different master. Leave is just reset but with keeping the system ready to re-run `kubeadm join`. More importantly to me, reset needs to cordon and drain.

**Comment by luxas** [comment]
To me it sounds like you want: 
 - reset to do cordon and drain (which I'm a huge +1 on, I just haven't had that extra time)
 - leave to just run `reset` and `systemctl start kubelet`

And it sounds like `leave` then is unnecessary (unless we decide to rename `reset` to `leave` indeed)
We could consider `reset` to run `systemctl start kubelet` as well, but I think we should prefer to have as little interaction with systemd as possible. In fact, I would be more in favor for automatically trying to start the kubelet in that case.

**Comment by errordeveloper** [comment]
I agree that 'leave' doesn't seem necessary. I would rather contemplate
'join --reset'.

**Comment by luxas** [comment]
@errordeveloper I still want the top-level command, but I think we can just do fine with `reset` (or `leave` if we want to call it that)

**Comment by pires** [comment]
Feel free to reassign this to anyone else.

**Comment by kad** [comment]
```reset``` is applicable for both master and slave nodes, where ```leave``` mostly for non-master nodes (where graceful shutdown and removing from cluster is needed)

**Comment by luxas** [comment]
The PR is here: https://github.com/kubernetes/kubernetes/pull/37831

**Comment by luxas** [comment]
Fixed with kubernetes/kubernetes#37568

**Comment by cheburakshu** [comment]
@luxas `kubeadm reset` does not delete the network interfaces. It does delete the /etc/cni/net.d though. Should it not do a complete reset? I did raise an issue with weavenet [here](https://github.com/weaveworks/weave/issues/2911), but it is more of a kubeadm cleanup issue.

**Comment by luxas** [comment]
@cheburakshu Cleaning up the network is hard; you're free to install whatever networking solution you want, so therefore kubeadm can't know how to cleanup something it hasn't installed.

Unfortunately this is out of scope for kubeadm, we can't build in cleanup solutions for every single network provider.
